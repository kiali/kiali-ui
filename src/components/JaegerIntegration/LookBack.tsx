import * as React from 'react';
import { FormSelect, FormSelectOption } from '@patternfly/react-core';
import { config, serverConfig } from '../../config';
import { style } from 'typestyle';

interface LookBackProps {
  disabled?: boolean;
  lookback?: number;
  setLookback: (value: string, event: any) => void;
}

const lookbackDropdown = style({ marginLeft: '-80px' });

export class LookBack extends React.PureComponent<LookBackProps> {
  lookBackOptions = { ...serverConfig.durations, ...{ 0: 'Custom Time Range' } };
  lookbackDefault = config.toolbar.defaultDuration;

  componentDidMount() {
    this.props.setLookback(String(this.props.lookback), null);
  }

  render() {
    const { lookback, setLookback } = this.props;
    const options: any[] = [];
    for (const [key, value] of Object.entries(this.lookBackOptions)) {
      options.push({ value: key, label: value, disabled: false });
    }

    return (
      <>
        <FormSelect
          value={lookback !== undefined ? lookback : this.lookbackDefault}
          onChange={setLookback}
          aria-label="FormSelect lookback"
          className={lookbackDropdown}
        >
          {options.map((option, index) => (
            <FormSelectOption isDisabled={option.disabled} key={index} value={option.value} label={option.label} />
          ))}
        </FormSelect>
      </>
    );
  }
}

export default LookBack;
