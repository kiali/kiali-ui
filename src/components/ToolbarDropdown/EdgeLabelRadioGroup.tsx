import * as React from 'react';
import PropTypes from 'prop-types';
import { style } from 'typestyle';
import * as _ from 'lodash';

import { GraphParamsType } from '../../types/Graph';
import { EdgeLabelMode } from '../../types/GraphFilter';

const EDGE_LABEL_MODES = EdgeLabelMode.getValues().reduce((map, edgeLabelMode) => {
  map[edgeLabelMode] = _.capitalize(_.startCase(edgeLabelMode));
  return map;
}, {});

type EdgeLabelRadioGroupProps = {
  graphParams: GraphParamsType;
  onEdgeChanged: PropTypes.func;
};

export class EdgeLabelRadioGroup extends React.Component<EdgeLabelRadioGroupProps> {
  constructor(props: EdgeLabelRadioGroupProps) {
    super(props);
  }

  onEdgeChanged = (event: any) => {
    const selectedEdge = event.target.value;
    this.setState({
      edgeValue: selectedEdge
    });
    this.props.onEdgeChanged(event);
  };

  render() {
    const radioButtonStyle = style({ marginLeft: 5 });

    const edgeItems = Object.keys(EDGE_LABEL_MODES).map((edgeLabelModeKey: any) => (
      <div key={edgeLabelModeKey}>
        <label className={radioButtonStyle}>
          <input
            type="radio"
            name="edge-label-radio-group"
            value={edgeLabelModeKey}
            checked={edgeLabelModeKey === this.props.graphParams.edgeLabelMode}
            onChange={this.onEdgeChanged}
          />
          <span className={radioButtonStyle}>{EDGE_LABEL_MODES[edgeLabelModeKey]}</span>
        </label>
      </div>
    ));

    return (
      <>
        <div>
          <label>Edge Labels:</label>
        </div>
        {edgeItems}
      </>
    );
  }
}

export default EdgeLabelRadioGroup;
