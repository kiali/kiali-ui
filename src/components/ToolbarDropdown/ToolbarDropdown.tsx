import * as React from 'react';
import { DropdownButton, MenuItem } from 'patternfly-react';
import PropTypes from 'prop-types';

type ToolbarDropdownProps = {
  disabled: boolean;
  nameDropdown?: string;
  initialValue: number | string;
  initialLabel: string | undefined;
  onClick: PropTypes.func;
  options: PropTypes.Map<string | number, string>;
};

type ToolbarDropdownState = {
  currentValue: number | string;
  currentName: string | undefined;
};

export class ToolbarDropdown extends React.Component<ToolbarDropdownProps, ToolbarDropdownState> {
  constructor(props: ToolbarDropdownProps) {
    super(props);
    this.state = {
      currentValue: props.initialValue,
      currentName: props.initialLabel
    };
  }

  onKeyChanged = (key: any) => {
    this.setState({ currentValue: key[0], currentName: key[1] });
    this.props.onClick(key[0]);
  };

  render() {
    return (
      <div className="form-group">
        {this.props.nameDropdown && <label style={{ paddingRight: '0.5em' }}>{this.props.nameDropdown}:</label>}
        <DropdownButton title={this.state.currentName} onSelect={this.onKeyChanged}>
          {this.props.options.map(r => (
            <MenuItem key={r[0]} active={r[0] === this.state.currentValue} eventKey={r}>
              {r[1]}
            </MenuItem>
          ))}
        </DropdownButton>
      </div>
    );
  }
}
