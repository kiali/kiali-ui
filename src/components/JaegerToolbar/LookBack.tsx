import * as React from 'react';
import { connect } from 'react-redux';
import { Col, Form, FormGroup, FormControl, FieldLevelHelp } from 'patternfly-react';
import { KialiAppState } from '../../store/Store';
import ToolbarDropdown from '../../components/ToolbarDropdown/ToolbarDropdown';
import { JaegerActions } from '../../actions/JaegerActions';
import { ThunkDispatch } from 'redux-thunk';
import { KialiAppAction } from '../../actions/KialiAppAction';

interface LookBackProps {
  fetching: boolean;
  setLookback: (lookback: string) => void;
  lookback: string;
  onChangeCustom: (when: string, dateField: string, timeField: string) => void;
}

export class LookBack extends React.PureComponent<LookBackProps, {}> {
  lookBackOptions = {
    '1h': 'Last Hour',
    '2h': 'Last 2 Hours',
    '3h': 'Last 3 Hours',
    '6h': 'Last 6 Hours',
    '12h': 'Last 12 Hours',
    '24h': 'Last 24 Hours',
    '2d': 'Last 2 Days',
    custom: 'Custom Time Range'
  };

  constructor(props: LookBackProps) {
    super(props);
  }

  componentDidMount() {
    this.props.setLookback(this.props.lookback);
  }

  render() {
    const { lookback, fetching, setLookback, onChangeCustom } = this.props;
    const tz = lookback === 'custom' ? new Date().toTimeString().replace(/^.*?GMT/, 'UTC') : null;

    return (
      <span style={{ marginLeft: '10px' }}>
        <Col componentClass={Form.ControlLabel} style={{ marginRight: '10px' }}>
          Lookback
        </Col>
        <ToolbarDropdown
          id="lookback-selector"
          disabled={fetching}
          options={this.lookBackOptions}
          value={lookback}
          label={this.lookBackOptions[lookback]}
          useName={false}
          handleSelect={setLookback}
        />
        {tz && (
          <Form style={{ display: 'inline-flex' }} inline={true}>
            <FormGroup>
              <Col componentClass={Form.ControlLabel}>
                Start Time
                <FieldLevelHelp
                  style={{ marginLeft: '10px' }}
                  content={<div>Times are expressed in {tz}</div>}
                  placement={'bottom'}
                />
              </Col>

              <FormControl
                style={{ marginLeft: '10px' }}
                type="date"
                disabled={false}
                onChange={e => onChangeCustom('start', e.target.value, '')}
              />
              <FormControl
                style={{ marginLeft: '10px' }}
                type="time"
                disabled={false}
                onChange={e => onChangeCustom('start', '', e.target.value)}
              />
            </FormGroup>
            <FormGroup>
              <Col componentClass={Form.ControlLabel}>
                End Time
                <FieldLevelHelp
                  style={{ marginLeft: '10px' }}
                  content={<div>Times are expressed in {tz}</div>}
                  placement={'bottom'}
                />
              </Col>

              <FormControl
                style={{ marginLeft: '10px' }}
                type="date"
                disabled={false}
                onChange={e => onChangeCustom('end', e.target.value, '')}
              />
              <FormControl
                style={{ marginLeft: '10px' }}
                type="time"
                disabled={false}
                onChange={e => onChangeCustom('end', '', e.target.value)}
              />
            </FormGroup>
          </Form>
        )}
      </span>
    );
  }
}

const mapStateToProps = (state: KialiAppState) => {
  return {
    fetching: state.jaegerState.search.serviceSelected === '',
    lookback: state.jaegerState.search.lookback
  };
};

const mapDispatchToProps = (dispatch: ThunkDispatch<KialiAppState, void, KialiAppAction>) => {
  return {
    setLookback: (lookback: string) => {
      dispatch(JaegerActions.setLookback(lookback));
    }
  };
};

const LookBackContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(LookBack);

export default LookBackContainer;
