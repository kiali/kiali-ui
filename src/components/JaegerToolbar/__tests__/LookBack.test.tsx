import * as React from 'react';
import { shallow } from 'enzyme';
import { LookBack } from '../LookBack';
import { Form, FormGroup } from 'patternfly-react';
import ToolbarDropdown from '../../../components/ToolbarDropdown/ToolbarDropdown';

const lookBackOptions = {
  '1h': 'Last Hour',
  '2h': 'Last 2 Hours',
  '3h': 'Last 3 Hours',
  '6h': 'Last 6 Hours',
  '12h': 'Last 12 Hours',
  '24h': 'Last 24 Hours',
  '2d': 'Last 2 Days',
  custom: 'Custom Time Range'
};

describe('LookBack', () => {
  let wrapper, onChangeCustom, setLookback;

  beforeEach(() => {
    onChangeCustom = jest.fn();
    setLookback = jest.fn();
    wrapper = shallow(
      <LookBack onChangeCustom={onChangeCustom} setLookback={setLookback} fetching={false} lookback={''} />
    );
  });

  it('renders LookBack correctly without custom', () => {
    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });

  it('renders Extra forms when lookback is custom', () => {
    expect(wrapper.find(Form).length).toEqual(0);
    wrapper.setProps({ lookback: 'custom' });
    expect(wrapper.find(Form).length).toEqual(1);
    expect(wrapper.find(FormGroup).length).toEqual(2);
  });

  it('LookBack has lookBackOptions options', () => {
    expect(wrapper.instance().lookBackOptions).toEqual(lookBackOptions);
  });

  it('disable ToolbarDropwdown if no namespaces', () => {
    expect(wrapper.find(ToolbarDropdown).length).toEqual(1);
    expect(
      wrapper
        .find(ToolbarDropdown)
        .first()
        .props()['disabled']
    ).toBeFalsy();
    wrapper.setProps({ fetching: true });
    expect(
      wrapper
        .find(ToolbarDropdown)
        .first()
        .props()['disabled']
    ).toBeTruthy();
  });

  it('ToolbarDropdown has  lookBackOptions like options', () => {
    expect(
      wrapper
        .find(ToolbarDropdown)
        .first()
        .props()['options']
    ).toEqual(lookBackOptions);
  });

  it('ToolbarDropdown has 1h by defaut', () => {
    wrapper.setProps({ lookback: '1h' });
    expect(
      wrapper
        .find(ToolbarDropdown)
        .first()
        .props()['value']
    ).toEqual('1h');
  });

  it('ToolbarDropdown has setLookback like handleSelect', () => {
    expect(
      wrapper
        .find(ToolbarDropdown)
        .first()
        .props()['handleSelect']
    ).toBe(setLookback);
  });
});
