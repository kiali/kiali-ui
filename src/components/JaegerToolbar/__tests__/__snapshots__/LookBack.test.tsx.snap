// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LookBack renders LookBack correctly without custom 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LookBack
    fetching={false}
    lookback=""
    onChangeCustom={[MockFunction]}
    setLookback={
      [MockFunction] {
        "calls": Array [
          Array [
            "",
          ],
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Col
          bsClass="col"
          componentClass={[Function]}
          style={
            Object {
              "marginRight": "10px",
            }
          }
        >
          Lookback
        </Col>,
        <ToolbarDropdown
          disabled={false}
          handleSelect={
            [MockFunction] {
              "calls": Array [
                Array [
                  "",
                ],
              ],
            }
          }
          id="lookback-selector"
          label={undefined}
          options={
            Object {
              "12h": "Last 12 Hours",
              "1h": "Last Hour",
              "24h": "Last 24 Hours",
              "2d": "Last 2 Days",
              "2h": "Last 2 Hours",
              "3h": "Last 3 Hours",
              "6h": "Last 6 Hours",
              "custom": "Custom Time Range",
            }
          }
          useName={false}
          value=""
        />,
        null,
      ],
      "style": Object {
        "marginLeft": "10px",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "bsClass": "col",
          "children": "Lookback",
          "componentClass": [Function],
          "style": Object {
            "marginRight": "10px",
          },
        },
        "ref": null,
        "rendered": "Lookback",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "disabled": false,
          "handleSelect": [MockFunction] {
            "calls": Array [
              Array [
                "",
              ],
            ],
          },
          "id": "lookback-selector",
          "label": undefined,
          "options": Object {
            "12h": "Last 12 Hours",
            "1h": "Last Hour",
            "24h": "Last 24 Hours",
            "2d": "Last 2 Days",
            "2h": "Last 2 Hours",
            "3h": "Last 3 Hours",
            "6h": "Last 6 Hours",
            "custom": "Custom Time Range",
          },
          "useName": false,
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      null,
    ],
    "type": "span",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Col
            bsClass="col"
            componentClass={[Function]}
            style={
              Object {
                "marginRight": "10px",
              }
            }
          >
            Lookback
          </Col>,
          <ToolbarDropdown
            disabled={false}
            handleSelect={
              [MockFunction] {
                "calls": Array [
                  Array [
                    "",
                  ],
                ],
              }
            }
            id="lookback-selector"
            label={undefined}
            options={
              Object {
                "12h": "Last 12 Hours",
                "1h": "Last Hour",
                "24h": "Last 24 Hours",
                "2d": "Last 2 Days",
                "2h": "Last 2 Hours",
                "3h": "Last 3 Hours",
                "6h": "Last 6 Hours",
                "custom": "Custom Time Range",
              }
            }
            useName={false}
            value=""
          />,
          null,
        ],
        "style": Object {
          "marginLeft": "10px",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bsClass": "col",
            "children": "Lookback",
            "componentClass": [Function],
            "style": Object {
              "marginRight": "10px",
            },
          },
          "ref": null,
          "rendered": "Lookback",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "disabled": false,
            "handleSelect": [MockFunction] {
              "calls": Array [
                Array [
                  "",
                ],
              ],
            },
            "id": "lookback-selector",
            "label": undefined,
            "options": Object {
              "12h": "Last 12 Hours",
              "1h": "Last Hour",
              "24h": "Last 24 Hours",
              "2d": "Last 2 Days",
              "2h": "Last 2 Hours",
              "3h": "Last 3 Hours",
              "6h": "Last 6 Hours",
              "custom": "Custom Time Range",
            },
            "useName": false,
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        null,
      ],
      "type": "span",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
